lemur/plugins/lemur_gcp/endpoints.py

```diff
@@ -0,0 +1,168 @@
+from collections import defaultdict
+from google.cloud.compute_v1.services import ssl_certificates, ssl_policies, \
+    global_forwarding_rules, target_https_proxies, target_ssl_proxies
+from google.cloud.compute_v1 import TargetHttpsProxiesSetSslCertificatesRequest, \
+    TargetSslProxiesSetSslCertificatesRequest
+
+from lemur.plugins.lemur_gcp import certificates
+
+
+def fetch_target_proxies(project_id, credentials):
+    """
+    Fetches HTTPs target proxies for L7 traffic and SSL target proxies for L4 traffic.
+
+    :param project_id:
+    :param credentials:
+    :return:
+    """
+    endpoints = []
+    forwarding_rules_map = fetch_global_forwarding_rules_map(project_id, credentials)
+    http_proxies_client = target_https_proxies.TargetHttpsProxiesClient(credentials=credentials)
+    ssl_policies_client = ssl_policies.SslPoliciesClient(credentials=credentials)
+    ssl_client = ssl_certificates.SslCertificatesClient(credentials=credentials)
+    for proxy in http_proxies_client.list(project=project_id):
+        endpoint = get_endpoint_from_proxy(project_id, proxy, ssl_policies_client, ssl_client, forwarding_rules_map)
+        if endpoint:
+            endpoints.append(endpoint)
+    ssl_proxies_client = target_ssl_proxies.TargetSslProxiesClient(credentials=credentials)
+    for proxy in ssl_proxies_client.list(project=project_id):
+        endpoint = get_endpoint_from_proxy(project_id, proxy, ssl_policies_client, ssl_client, forwarding_rules_map)
+        if endpoint:
+            endpoints.append(endpoint)
+    return endpoints
+
+
+def get_endpoint_from_proxy(project_id, proxy, ssl_policies_client, ssl_client, forwarding_rules_map):
+    """
+    Converts a proxy (either HTTPs or SSL) into a Lemur endpoint.
+    :param project_id:
+    :param proxy:
+    :param ssl_policies_client:
+    :param ssl_client:
+    :param forwarding_rules_map:
+    :return:
+    """
+    kind = proxy.kind.split("#")[-1].lower()
+    if kind not in ("targethttpsproxy", "targetsslproxy"):
+        return None
+    if len(proxy.ssl_certificates) == 0:
+        return None
+    fw_rules = forwarding_rules_map.get(proxy.self_link, None)
+    if not fw_rules:
+        return None
+    # The first certificate is the primary.
+    # See https://cloud.google.com/sdk/gcloud/reference/compute/target-https-proxies/update
+    ssl_cert_name = get_name_from_self_link(proxy.ssl_certificates[0])
+    cert = ssl_client.get(project=project_id, ssl_certificate=ssl_cert_name)
+    primary_certificate = dict(
+        name=cert.name,
+        registry_type="gcp",
+        path="",
+    )
+    fw_rule_ingresses = []
+    for rule in fw_rules:
+        ip = rule.I_p_address
+        port = rule.port_range.split("-")[0]
+        fw_rule_ingresses.append((ip, port))
+    fw_rule_ingresses.sort()
+    primary_ip, primary_port = fw_rule_ingresses[0][0], fw_rule_ingresses[0][1]
```

Comment:
Do you mind going into a bit of detail about what is the logic for line 62-68.
If Im understanding correctly the for loop will create an array of forwarding rules for the endpoint. So for an GCP LB forwarding both 80 and 443 our data structure may look something like
```[('11.12.13.14', '443'), ('11.12.13.14', '80')]```
We then `sort()` and take the IP and Port of the first element. 
Why are we preferring one over the other?

Reply:
I didn't want to assume anything about how the forwarding rules get returned so I tried to make it stable for each time Lemur runs a full sync of the endpoints (assuming a FW rule doesn't get added in between syncs)

Reply:
Ah understood!